meta-client/src/main/java/org/apache/helix/metaclient/recipes/leaderelection/LeaderElectionClient.java

```diff
@@ -106,23 +182,65 @@ public boolean joinLeaderElectionParticipantPool(String leaderPath, Object userI
+   * @throws RuntimeException If the participant did not join participant pool via this client.
    */
-  public boolean exitLeaderElectionParticipantPool(String leaderPath) {
-    return false;
+  public void exitLeaderElectionParticipantPool(String leaderPath) {
+    _metaClient.unsubscribeDataChange(leaderPath + LEADER_ENTRY_KEY, _reElectListener);
+    // TODO: remove from pool folder
+    relinquishLeaderHelper(leaderPath, true);
   }
 
   /**
-   * Releases leadership for participant.
+   * Releases leadership for participant. Still stays in the participant pool.
    *
    * @param leaderPath The path for leader election.
    *
    * @throws RuntimeException if the leadership is not owned by this participant, or if the
-   *                          participant did not join participant pool via this client. // TODO: define exp type
+   *                          participant did not join participant pool via this client.
    */
   public void relinquishLeader(String leaderPath) {
+    relinquishLeaderHelper(leaderPath, false);
+  }
+
+  /**
+   * relinquishLeaderHelper and LeaderElectionParticipantPool if configured
+   * @param leaderPath
+   * @param exitLeaderElectionParticipantPool
+   */
+  private void relinquishLeaderHelper(String leaderPath, Boolean exitLeaderElectionParticipantPool) {
+    String key = leaderPath + LEADER_ENTRY_KEY;
+    // if current client is in the group
+    if (leaderGroups.contains(key)) {
+      // remove leader path from leaderGroups after check if exiting the pool.
+      // to prevent a race condition in In Zk implementation:
+      // If there are delays in ZkClient event queue, it is possible the leader election client received leader
+      // deleted event after unsubscribeDataChange. We will need to remove it from in memory `leaderGroups` map before
+      // deleting ZNode. So that handler in ReElectListener won't recreate the leader node.
+      if (exitLeaderElectionParticipantPool) {
+        leaderGroups.remove(leaderPath + LEADER_ENTRY_KEY);
+      }
+      // check if current participant is the leader
+      // read data and stats, check, and multi check + delete
+      ImmutablePair<LeaderInfo, MetaClientInterface.Stat> tup = _metaClient.getDataAndStat(key);
+      if (tup.left.getLeaderName().equalsIgnoreCase(_participant)) {
+        int expectedVersion = tup.right.getVersion();
+        List<Op> ops = Arrays.asList(Op.check(key, expectedVersion), Op.delete(key, expectedVersion));
+        //Execute transactional support on operations
+        List<OpResult> opResults = _metaClient.transactionOP(ops);
+        if (opResults.get(0).getType() == ERRORRESULT) {
+          if (isLeader(leaderPath)) {
+            // Participant re-elected as leader.
+            throw new ConcurrentModificationException("Concurrent operation, please retry");
+          } else {
+            LOG.info("Someone else is already leader");
+          }
+        }
+      }
+    } else {
```

Comment:
it may be better to have reverse condition, so you exit cleanly first.. ie. reverse the if (condition), so that this code exits first.